src/csharp/Grpc.Core.Api/AsyncStreamReaderExtensions.cs.cs

```diff
@@ -0,0 +1,47 @@
+#region Copyright notice and license
+
+// Copyright 2015 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Grpc.Core
+{
+#if GRPC_CSHARP_SUPPORT_ASYNC_ENUMERABLE
+    /// <summary>
+    /// Extension methods for <see cref="IAsyncStreamReader{T}"/>.
+    /// </summary>
+    public static class IAsyncStreamReaderExtensions
+    {
+        /// <summary>
+        /// Read all results from the specified stream reader.
+        /// </summary>
+        public async static IAsyncEnumerable<T> ReadAllAsync<T>(this IAsyncStreamReader<T> streamReader, CancellationToken cancellationToken = default)
```

Comment:
The cancellation token should have an [EnumeratorCancellation] attribute. This allows the token to flow through from GetAsyncEnumerator(token) to this method.